exodus_gw/aws/util.py

```diff
@@ -1,15 +1,47 @@
 import io
 import logging
 import re
-from typing import AnyStr
+from typing import AnyStr, Dict
 from xml.etree.ElementTree import Element, ElementTree, SubElement
 
 from defusedxml.ElementTree import fromstring
-from fastapi import HTTPException, Response
+from fastapi import HTTPException, Request, Response
+
+from ..settings import Settings
 
 LOG = logging.getLogger("exodus-gw")
 
 
+def extract_request_metadata(request: Request, settings: Settings):
+    # Any headers prefixed with "x-amz-meta-" will be picked out as
+    # metadata for s3 upload.
+    #
+    # https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingMetadata.html
+    metadata = {}
+    for k, v in request.headers.items():
+        if k.startswith("x-amz-meta-"):
+            metadata[k.replace("x-amz-meta-", "")] = v
```

Comment:
This is slightly incorrect since it will replace *all* instances of x-amz-meta-, not just the leading prefix.  Though it would be very weird to include an x-amz-meta- inside the metadata field name, it would be legal to do so and we shouldn't silently strip it out if so.

It can be fixed to strip just the leading prefix like this:

```suggestion
            metadata[k.replace("x-amz-meta-", "", 1)] = v
```